src/protocols/secure_channel/CASESession.h

```diff
@@ -211,6 +190,119 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
     // If this function returns true, the CASE session has been reset and is ready for a new session establishment.
     bool InvokeBackgroundWorkWatchdog();
 
+protected:
+    // Helper Enum for use in HandleSigma1_and_SendSigma2
+    enum class Step : uint8_t
+    {
+        kNone,
+        kSendSigma2,
+        kSendSigma2Resume,
+    };
+    // Making NextStep a Variant allows HandleSigma() to return either a Step value (indicating
+    // the next Sigma step to send) or a CHIP_ERROR (indicating a failure that will trigger
+    // a Status Report).
+    using NextStep = Variant<Step, CHIP_ERROR>;
+    // This struct  only serves as a base struct for EncodedSigma1Inputs and ParsedSigma1
+    struct Sigma1Param
+    {
+        ByteSpan initiatorRandom;
+        uint16_t initiatorSessionId;
+        ByteSpan destinationId;
+        bool sessionResumptionRequested = false;
+        ByteSpan resumptionId;
+        ByteSpan initiatorResumeMIC;
+    };
+
+    struct EncodeSigma1Inputs : Sigma1Param
+    {
+        const Crypto::P256PublicKey * initiatorEphPubKey         = nullptr;
+        const ReliableMessageProtocolConfig * initiatorMrpConfig = nullptr;
+        uint8_t initiatorResume1MICBuffer[Crypto::CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES];
+    };
+
+    struct ParsedSigma1 : Sigma1Param
+    {
+        ByteSpan initiatorEphPubKey;
+        bool initiatorMrpParamsPresent = false;
+    };
+
+    struct EncodeSigma2Inputs
+    {
+        uint8_t responderRandom[kSigmaParamRandomNumberSize];
+        uint16_t responderSessionId;
+        const Crypto::P256PublicKey * responderEphPubKey = nullptr;
+        // ScopedMemoryBufferWithSize is not used for msgR2Encrypted since encrypted2Length might differ from the allocated buffer
+        // size
+        Platform::ScopedMemoryBuffer<uint8_t> msgR2Encrypted;
+        size_t encrypted2Length = 0;
+        const ReliableMessageProtocolConfig * responderMrpConfig;
+    };
+
+    struct EncodeSigma2ResumeInputs
+    {
+        ByteSpan resumptionId;
+        uint8_t sigma2ResumeMICBuffer[Crypto::CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES];
+        MutableByteSpan resumeMICSpan{ sigma2ResumeMICBuffer };
+        uint16_t responderSessionId;
+        const ReliableMessageProtocolConfig * responderMrpConfig;
+    };
+
+    /**
+     * @brief  Encodes a Sigma1 message into TLV format and allocates a buffer for it, which is owned by the PacketBufferHandle
+     *outparam.
+     *
+     * @param outMsg     PacketBufferHandle passed by reference. A new buffer will be allocated and assigned to it within the
+     *method.
+     *
+     * @param inParam a struct containing all the values that will be encoded into TLV format
+     *
+     **/
+    static CHIP_ERROR EncodeSigma1(System::PacketBufferHandle & outMsg, EncodeSigma1Inputs & inParam);
+
+    /**
+     * Parse a sigma1 message.  This function will return success only if the
```

Comment:
```suggestion
     * Parse a Sigma1 message.  This function will return success only if the
```
This was https://github.com/project-chip/connectedhomeip/pull/36679#discussion_r1889498267

Reply:
fixed for real now